### ISSUE TITLE
prevent non-owner user from editing/deleting the snippet

### ISSUE THREAD

**Issue by ahmednooor** [bartTC/dpaste#98]
hi,
i was able to delete someone else's snippet. i think only the owner of the snippet should be able to delete/edit the snippet. since there is no login system (which i understand as a way of keeping things simple), how about giving a unique key (can be random chars) along with the URL of the snippet to the user at the time of snippet creation and ask them to save that URL and KEY. every time someone wants to edit/delete snippet they should be asked to provide the secret key. (just a suggestion)
thanks.

**Comment by mzmcbride** [comment]
When you create a snippet, it automatically gets assigned an ID; for example, `rCZC` or `ZNko`. This string is treated as both the snippet ID and as a secret ID in the code. As noted at <https://dpaste.de/about/>, this ID can be used to delete any paste ("Actually you can delete any snippet of anybody, as long as you know the short code."); source:
 <https://github.com/bartTC/dpaste/blob/b8c8cb79/dpaste/templates/dpaste/about.html#L78-L79>.

It's not clear how the unique key of random characters that you're proposing in this issue would be different from what's already been implemented at <https://github.com/bartTC/dpaste/blob/b8c8cb79/dpaste/models.py#L13>.

**Comment by ahmednooor** [comment]
@mzmcbride  i am proposing that a unique secret key different from URL is provided to the owner only once at the time of creation of the snippet so only the owner may edit/delete the snippet. 
e.g. when u assign a unique code to the snippet URL, create another unique code different from the one in URL and show it to the user when he/she first saves the snippet and then every time someone wants to edit/delete the snippet, the app asks for the other secret key provided to the owner at the time of snippet creation. this way only the person with the secret key provided at the time of creation can edit/delete the snippet. i am proposing this in the context of web where anybody can delete any body's snippet. i accidentally deleted a person's snippet who was asking for some code help that is why i think only the owner should be able to edit/delete the snippet.

**Comment by wanderleihuttel** [comment]
I noticed the same issue. 
I have almost sure that in the previous versions only the owner could edit or delete the snippet.

**Comment by bartTC** [comment]
>  I have almost sure that in the previous versions only the owner could edit or delete the snippet.

You're right. I intentionally removed this so people can delete any snippet at any time. I had some issues with fraud in the past, where people put stolen credentials on dpaste.de and I wanted to give anyone the opportunity to delete them asap.

I might bring it back.

**Comment by bartTC** [comment]
> i am proposing that a unique secret key different from URL is provided to the owner only once at the time of creation of the snippet so only the owner may edit/delete the snippet.

This isn't realistic. The reason for the form on the about page is that people who put on a snippet via the API can delete it. Nobody would write down a special 'delete' code manually.

**Comment by ahmednooor** [comment]
Perhaps storing the sec key in `localStorage`? So only the device and browser through which the snippet was created, will be able to delete the snippet? And not anybody will be able to delete anybody's snippet. The user won't have to remember or note down any key.
Though there might be cases where the user clears the cache and loose the `localStorage`, in that case the snippet will just expire after the expiry date.

This is just a suggestion. Feel free to close this issue.

**Comment by bartTC** [comment]
@ahmednooor I think it makes sense to have this, but the reason for this form and the actual 'free' delete action was that people which created a snippet via the API couldn't delete the snippet. There's no cookie on local storage if you create a snippet with the API.

**Comment by ahmednooor** [comment]
IMO most common users use dpaste.de to use the service. The point of API would be to accommodate the desktop client right? well the maintainer of the desktop client can store the sec key in sqlite db. 

Snippet Creation:
1) client sends a request to store a snippet to the api.
2) the api at the backend creates a random string for link and creates another random string(sec key) and store both in the same field in db.
3) api returns both the link and the sec key to the client so that the client can store both.

Snippet Deletion:
1) the client sends the link and the stored sec key to the api.
2) the api matches both and checks if the sec key is associated with the link.
3) if it checks out the api deletes the snippet and returns true and status msg. if it not checks out it returns false and error msg.
e.g.  {"status":true, "msg":"snippet deleted"}  or  {"status":false,"msg":"sec key not matched etc."}

This way the sec key stuff will be the responsibility of the maintainer and not the common user.

BTW it is not that much of a problem. Not everybody deletes someone's snippet due to bad intention. It mostly happens mistakenly. So feel free to leave it as it is if it requires too much code changing in the core api. A working code is always better than the hypothetical one.

Feel free to close this issue.
